docs/proposals/container-init.md

```diff
@@ -0,0 +1,333 @@
+according to the pod RestartPolicy. RestartPolicyNever pods will immediately
+fail and exit. RestartPolicyAlways pods will retry the failing init container
+with increasing backoff until it succeeds. Future revisions of this spec may
+add the ability for an init container to define a limited container restart
+policy distinct from the pod RestartPolicy.
+
+A pod cannot be ready until all init containers have succeeded. The ports
+on an init container are not aggregated under a service. A pod that is
+being initialized is in the `Pending` phase but should have a distinct
+condition.
+
+If the pod is "restarted" all containers stopped and started due to
+a node restart, change to the pod definition, or admin interaction, all
+init containers must execute again.
+
+Each init container has all of the fields of an app container. The following
+fields are prohibited from being used on init containers by validation:
+
+* `readinessProbe` - init containers must exit for pod startup to continue,
+  are not included in rotation, and so cannot define readiness distinct from
+  completion.
+
+Because init containers are semantically different in lifecycle from app
+containers (they are run serially, rather than in parallel), for backwards
+compatibility with existing clients they must be identified as distinct
+fields in the API:
+
+    pod:
+      spec:
+        containers: ...
+        initContainers:
+        - name: init-container1
+          image: ...
+          ...
+        - name: init-container2
+        ...
+      status:
+        containerStatuses: ...
+        initContainerStatuses:
+        - name: init-container1
+          ...
+        - name: init-container2
+          ...
+
+This separation also serves to make the order of container initialization
+clear - init containers are executed in the order that they appear, then all
+app containers are started at once.
+
+The name of each app and init container in a pod must be unique - it is a
+validation error for any container to share a name.
+
+
+### Resources
+
+Given the ordering and execution for init containers, the following rules
+for resource usage apply:
+
+* The highest of any particular resource request or limit defined on all init
+  containers is the **effective init request/limit**
+* The pod's **effective request/limit** for a resource is the higher of:
+  * sum of all app containers request/limit for a resource
+  * effective init request/limit for a resource
+* The highest QoS tier of init containers is the **effective init QoS tier**
+
+So the following pod:
+
+    pod:
+      spec:
+        initContainers:
+        - limits:
+				    cpu: 100m
+				    memory: 1GiB
+        - limits:
+				    cpu: 50m
+				    memory: 2GiB
+			  containers:
+        - limits:
+				    cpu: 10m
+				    memory: 1100MiB
+        - limits:
+				    cpu: 10m
+				    memory: 1100MiB
+
+has an effective pod limit of `cpu: 100m`, `memory: 2200MiB` (highest init
```

Comment:
We can set the pod-level cgroups limits to match this value. If the difference between init container's and app container's resource requests are high, then it will lead to poor utilization though.